x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -5,75 +5,341 @@
+    private volatile long currentIndexMetadataVersion = 0;
+    private final AtomicInteger retryCounter = new AtomicInteger(0);
+    private final Queue<Translog.Operation> buffer = new PriorityQueue<>(Comparator.comparing(Translog.Operation::seqNo).reversed());
+
+    ShardFollowNodeTask(long id, String type, String action, String description, TaskId parentTask, Map<String, String> headers,
+                        ShardFollowTask params, BiConsumer<TimeValue, Runnable> scheduler, TimeValue idleShardChangesRequestDelay,
+                        TimeValue retryTimeout) {
         super(id, type, action, description, parentTask, headers);
+        this.params = params;
+        this.scheduler = scheduler;
+        this.retryTimeout = retryTimeout;
+        this.idleShardChangesRequestDelay = idleShardChangesRequestDelay;
+    }
+
+    void start(long followGlobalCheckpoint) {
+        this.lastRequestedSeqno = followGlobalCheckpoint;
+        this.processedGlobalCheckpoint = followGlobalCheckpoint;
+        this.globalCheckpoint = followGlobalCheckpoint;
+
+        // Forcefully updates follower mapping, this gets us the leader imd version and
+        // makes sure that leader and follower mapping are identical.
+        updateMapping(imdVersion -> {
+            currentIndexMetadataVersion = imdVersion;
+            LOGGER.info("{} Started to follow leader shard {}, followGlobalCheckPoint={}, indexMetaDataVersion={}",
+                params.getFollowShardId(), params.getLeaderShardId(), followGlobalCheckpoint, imdVersion);
+            coordinateReads();
+        });
     }
 
+    private synchronized void coordinateReads() {
+        if (isStopped()) {
+            LOGGER.info("{} shard follow task has been stopped", params.getFollowShardId());
+            return;
+        }
+
+        LOGGER.trace("{} coordinate reads, lastRequestedSeqno={}, globalCheckpoint={}",
+            params.getFollowShardId(), lastRequestedSeqno, globalCheckpoint);
+        final int maxReadSize = params.getMaxBatchOperationCount();
+        while (hasReadBudget() && lastRequestedSeqno < globalCheckpoint) {
+            numConcurrentReads++;
+            long from = lastRequestedSeqno + 1;
+            LOGGER.trace("{}[{}] read [{}/{}]", params.getFollowShardId(), numConcurrentReads, from, maxReadSize);
+            long maxRequiredSeqno = Math.min(globalCheckpoint, from + maxReadSize);
+            sendShardChangesRequest(from, maxReadSize, maxRequiredSeqno);
+            lastRequestedSeqno = maxRequiredSeqno;
+        }
+
+        if (numConcurrentReads == 0) {
+            // We sneak peek if there is any thing new in the leader.
+            // If there is we will happily accept
+            numConcurrentReads++;
+            long from = lastRequestedSeqno + 1;
+            LOGGER.trace("{}[{}] peek read [{}]", params.getFollowShardId(), numConcurrentReads, from);
+            sendShardChangesRequest(from, maxReadSize, from);
```

Comment:
is this correct? don't we want  the maxrequired to be global checkpoint (i.e., from - 1 )? now it seem you have to bring at least on seq# back.